src/Dfc.CourseDirectory.Core/DataManagement/FileUploadProcessor.Apprenticeships.cs

```diff
@@ -411,6 +411,62 @@ internal async Task<(int[] Missing, (int StandardCode, int StandardVersion, int
             }
         }
 
+        private async Task AcquireExclusiveApprenticeshipUploadLockForProvider(Guid providerId, ISqlQueryDispatcher sqlQueryDispatcher)
+        {
+            var lockName = $"DM_Apprenticeship:{providerId}";
+            const int timeoutMilliseconds = 3000;
+
+            var acquired = await sqlQueryDispatcher.ExecuteQuery(new GetExclusiveLock()
+            {
+                Name = lockName,
+                TimeoutMilliseconds = timeoutMilliseconds
+            });
+
+            if (!acquired)
+            {
+                throw new Exception($"Failed to acquire exclusive apprenticeship upload lock for provider {providerId}.");
+            }
+        }
+
+        public async Task<(IReadOnlyCollection<ApprenticeshipUploadRow> Rows, UploadStatus UploadStatus)> GetApprenticeshipUploadRowsForProvider(Guid providerId)
+        {
+            using (var dispatcher = _sqlQueryDispatcherFactory.CreateDispatcher())
+            {
+                await AcquireExclusiveApprenticeshipUploadLockForProvider(providerId, dispatcher);
```

Comment:
Could do with adding this method call to all other places similar to courses and venues too

Reply:
I've added these now.